From 78cdd92f9852e6894fba8eda1ff7dd2325bdf0b3 Mon Sep 17 00:00:00 2001
From: Matt Robinson <git@nerdoftheherd.com>
Date: Sun, 4 Sep 2022 19:21:34 +0100
Subject: [PATCH] Make default compat shell paths configurable

Allow overriding the default shell paths returned by the compatibility
getusershell() function with new macros DROPBEAR_PATH_BSHELL and
DROPBEAR_PATH_CSHELL.  Default these to the values of _PATH_BSHELL and
_PATH_CSHELL if they are defined (generally in paths.h) or fall back to
the original sane values if neither is set.

This fixes server login under Android (without having to add
/etc/shells) and should provide flexibility for other systems missing
getusershell with odd file system layouts.
---
 compat.c |  2 +-
 compat.h | 15 +++++++++++++++
 2 files changed, 16 insertions(+), 1 deletion(-)

diff --git a/compat.c b/compat.c
index 7a0e78a..e1a92d2 100644
--- a/compat.c
+++ b/compat.c
@@ -232,7 +232,7 @@ void setusershell() {
 
 static char **initshells() {
 	/* don't touch this list. */
-	static const char *okshells[] = { "/bin/sh", "/bin/csh", NULL };
+	static const char *okshells[] = { DROPBEAR_PATH_BSHELL, DROPBEAR_PATH_CSHELL, NULL };
 	register char **sp, *cp;
 	register FILE *fp;
 	struct stat statb;
diff --git a/compat.h b/compat.h
index 58fd58e..ab683f0 100644
--- a/compat.h
+++ b/compat.h
@@ -47,6 +47,21 @@ char *basename(const char* path);
 char *getusershell(void);
 void setusershell(void);
 void endusershell(void);
+
+#ifndef DROPBEAR_PATH_BSHELL
+#ifdef _PATH_BSHELL
+#define DROPBEAR_PATH_BSHELL _PATH_BSHELL
+#else
+#define DROPBEAR_PATH_BSHELL "/bin/sh"
+#endif
+#endif
+#ifndef DROPBEAR_PATH_CSHELL
+#ifdef _PATH_CSHELL
+#define DROPBEAR_PATH_CSHELL _PATH_CSHELL
+#else
+#define DROPBEAR_PATH_CSHELL "/bin/csh"
+#endif
+#endif
 #endif
 
 #ifndef DROPBEAR_PATH_DEVNULL
-- 
2.37.3
